hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/keys/KeyStorage.java

```diff
@@ -0,0 +1,242 @@
+
+  private final Path privateKeyPath;
+  private final Path publicKeyPath;
+  private final KeyCodec keyCodec;
+  private final boolean externalKeysUsed;
+
+  /**
+   * Creates a KeyStorage object based on the configuration for the defined component, by assuming an empty key path
+   * suffix.
+   *
+   * @param config the SecurityConfiguration of the application
+   * @param component the component for which the KeyStorage is to be created
+   * @throws IOException in case the defined paths are unusable, or if the key algorithm in the configuration is
+   *                     unsupported.
+   * @see KeyStorage#KeyStorage(SecurityConfig, String, String) for more details.
+   */
+  public KeyStorage(SecurityConfig config, String component) throws IOException {
+    this(config, component, config.getKeyLocation(component));
+  }
+
+  /**
+   * Creates a KeyStorage object based on the configuration, the defined component, suffixing the base path with the
+   * defined suffix.<br/><br/>
+   * <p>
+   * The initialization is retrieving the base path from {@link SecurityConfig#getKeyLocation(String)} method,
+   * within which {@link SecurityConfig#getPrivateKeyFileName()} and {@link SecurityConfig#getPublicKeyFileName()}
+   * defined the name of the files that holds the public and private keys respectively.<br/>
+   * The base path is suffixed with the keyDirSuffix before resolving the path belongs to the key files in the folder.
+   * If the base path does not exists, it is created with "rwx------" POSIX permissions, if it exists, the code attempts
+   * to set its POSIX permissions to "rwx------". Key files are created with "rw-------" POSIX permission bits set.
+   * </p><br/>
+   * <p>
+   * In case {@link SecurityConfig#useExternalCACertificate(String)} returns true, the public and private key is read
+   * from {@link SecurityConfig#getExternalRootCaPublicKeyPath()} and
+   * {@link SecurityConfig#getExternalRootCaPrivateKeyPath()} respectively.
+   * If {@link SecurityConfig#useExternalCACertificate(String)} is true at the time of constructing this object, the
+   * store methods of the instance are not usable, and will throw an {@link UnsupportedOperationException}.
+   * </p>
+   * @param config the {@link SecurityConfig} of the system
+   * @param component the component name to be used
+   * @param keyDirSuffix the suffix to apply to the keys base path
+   * @throws IOException in case when base directory can not be created with the desired permissions, or if the key
+   *                     algorithm is not available for the {@link java.security.KeyFactory}
+   */
+  public KeyStorage(SecurityConfig config, String component, String keyDirSuffix) throws IOException {
+    this(config, component, Paths.get(config.getKeyLocation(component).toString() + keyDirSuffix));
+  }
+
+  private KeyStorage(SecurityConfig config, String component, Path keyPath) throws IOException {
+    if (config.useExternalCACertificate(component)) {
```

Comment:
@Galsza , previous implementation, it reads the public and private key from external directory, then write and save them to Ozone's directory.  In this new implementation, this rewrite is omitted. In this case, I would suggest check the permission of this external path too, make sure the path permission also comply with our requirement.

Reply:
@Galsza , thanks for updating the patch.  What I means is we should check the external directory permission, and enforce some rule. If we cannot enforce "rwx------", at least we should enforce that others group cannot have any access to this directory, and group cannot write to this directory,  for example "ozone.om.db.dirs.permissions". Because all the internal PKI system security is highly depends on the security of root private key. If anyone can access this external root certificate key materials, then the security will be more easy to compromise.